explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,153 @@
+
+* The binary **cannot** be stripped of its debug symbols.  `autofdo` does not support separate debug information files (i.e., it can't work with Ubuntu's `.ddeb` packages), and virtually all Ubuntu packages run `strip` doing their build in order to generate the `.ddeb` packages.  If you intend to profile an Ubuntu package, please keep that in mind.
+
+* The debug information file(s) **cannot** be processed by `dwz`.  This is a tool whose purpose is to compress the debug information generated when building a binary, and again, virtually all Ubuntu packages use it.  For this reason, it is currently not possible to profile most of Ubuntu's packages without first rebuilding them to disable `dwz` from running.
+
+* We have to be mindful of the options we pass to `perf`, particularly when it comes to recording branch prediction events.  The options will likely vary depending on whether you are using an Intel or AMD processor, for example.
+
+On top of that, the current `autofdo` version in Ubuntu (`0.19-3build3`, at the time of this writing) is not recent enough to process the `perfdata` files we will generate.  There is a PPA with a newer version of `autofdo` package for Ubuntu Noble [here](https://launchpad.net/~sergiodj/+archive/ubuntu/autofdo).  If you are running another version of Ubuntu and want to install a newer version of `autofdo`, you will need to build the software manually (please refer to the [upstream repository](https://github.com/google/autofdo) for further instructions).
+
+## A simple PGO scenario: `openssl speed`
+
+PGO makes sense when your software is CPU-bound, i.e., when it performs CPU intensive work and is not waiting on I/O, for example.  OpenSSL has a built-in benchmark command called `openssl speed`, which tests the performance of its cryptographic algorithms.  This is excellent for PGO because there is practically no I/O involved, and we are just constrained by how fast the CPU can run.
+
+### Running OpenSSL tests
+
+In order to measure the performance impact of PGO, it is important to perform several tests using an OpenSSL binary without PGO and then with it.
+
+The first thing we have to do is to do several runs of `openssl speed` *before* enabling PGO, so that we have a baseline to compare with.  As explained in the sections above, `autofdo` does not handle stripped binaries nor `dwz`-processed debug information, so we need to make
+After confirming that everything looks OK, we are ready to start the benchmark.  Let's run the command:
```

Comment:
This looks like it has eaten half of your sentence.
I assume you wanted to outline that you build it that way here?

Reply:
you have "Building OpenSSL for profiling" so I assume you want to refer to that here

Reply:
Argh, indeed.  Sorry about that.  Referred the next section and fixed the sentence.